File: pascal/functions_and_procedures/functions_overloading.pas
{$mode objfpc}

program FunctionsOverloading;

function some_function(i: integer): integer;
begin
  writeln('Function A arguments: ');
  writeln('i: ', i);
  some_function := 0;
end;

function some_function(r: real): integer;
begin
  writeln('Function B arguments: ');
  writeln('r: ', r);
  some_function := 0;
end;

function some_function(i: integer; r: real): integer;
begin
  writeln('Function C arguments: ');
  writeln('i: ', i);
  writeln('r: ', r);
  some_function := 0;
end;

function some_function(r: real; i: integer): integer;
begin
  writeln('Function D arguments: ');
  writeln('r: ', r);
  writeln('i: ', i);
  some_function := 0;
end;

function another_function(r: real): integer;
begin
  writeln('Function E arguments: ');
  writeln('r: ', r);
  another_function := 0;
end;

function another_function(i: integer; s: string = 'apple'): integer;
begin
  writeln('Function F arguments: ');
  writeln('i: ', i);
  writeln('s: ', s);
  another_function := 0;
end;

begin
  some_function(4.5);
  writeln();

  some_function(6);
  writeln();

  some_function(2, 1.5);
  writeln();

  some_function(3.6, 5);
  writeln();

  another_function(3);
  writeln();

  another_function(3.2);
  writeln();

  another_function(3, 'orange');
  writeln();
end.

File: c++/functions/functions_overloading.cpp
#include <iostream>

void some_function(int i)
{
  std::cout << "Function A arguments:\n"
    << "i: " << i << std::endl;
}

void some_function(double d)
{
  std::cout << "Function B arguments:\n"
    << "d: " << d << std::endl;
}

void some_function(int i, double d)
{
  std::cout << "Function C arguments:\n"
    << "i: " << i << std::endl
    << "d: " << d << std::endl;
}

void some_function(double d, int i)
{
  std::cout << "Function D arguments:\n"
    << "d: " << d << std::endl
    << "i: " << i << std::endl;
}

void another_function(double d)
{
  std::cout << "Function E arguments:\n"
    << "d: " << d << std::endl;
}

void another_function(int i, const char* s = "apple")
{
  std::cout << "Function F arguments:\n"
    << "i: " << i << std::endl
    << "s: " << s << std::endl;
}

int main()
{
  some_function(4.5);
  std::cout << std::endl;

  some_function(6);
  std::cout << std::endl;

  some_function(2, 1.5);
  std::cout << std::endl;

  some_function(3.6, 5);
  std::cout << std::endl;

  another_function(3);
  std::cout << std::endl;

  another_function(3.2);
  std::cout << std::endl;

  another_function(3.2, "orange");
  std::cout << std::endl;
}
